packages/beacon-node/src/util/execution.ts

```diff
@@ -8,10 +8,94 @@ import {IExecutionEngine} from "../execution/index.js";
 import {ChainEvent, ChainEventEmitter} from "../chain/emitter.js";
 import {BlockInputSource, IBlockInput} from "../chain/blocks/blockInput/types.js";
 import {Metrics} from "../metrics/index.js";
-import {fulu} from "@lodestar/types";
-import {isBlockInputColumns} from "../chain/blocks/blockInput/blockInput.js";
-import {ForkPostFulu} from "@lodestar/params";
-import {BLOB_AND_PROOF_V2_RPC_BYTES} from "../execution/engine/types.js";
+import {deneb, fulu} from "@lodestar/types";
+import {isBlockInputBlobs, isBlockInputColumns} from "../chain/blocks/blockInput/blockInput.js";
+import {ForkPostFulu, ForkPreFulu} from "@lodestar/params";
+import {computePreFuluKzgCommitmentsInclusionProof} from "./blobs.js";
+import {signedBlockToSignedHeader} from "@lodestar/state-transition";
+import {routes} from "@lodestar/api";
+import {toHex} from "@lodestar/utils";
+
+export async function getBlobSidecarsFromExecution(
+  config: ChainForkConfig,
+  executionEngine: IExecutionEngine,
+  metrics: Metrics | null,
+  emitter: ChainEventEmitter,
+  blockInput: IBlockInput
+) {
+  if (!isBlockInputBlobs(blockInput)) {
+    return;
+  }
+
+  if (blockInput.hasAllData()) {
+    return;
+  }
+
+  const forkName = blockInput.forkName as ForkPreFulu;
+  const blobMeta = blockInput.getMissingBlobMeta();
+
+  metrics?.blobs.getBlobsV1Requests.inc();
+  const enginedResponse = await executionEngine
+    .getBlobs(
+      forkName,
+      blobMeta.map(({versionedHash}) => versionedHash)
+    )
+    .catch((_e) => {
+      metrics?.blobs.getBlobsV1Error.inc();
+      return null;
+    });
+
+  if (enginedResponse === null) {
+    return;
+  }
+
+  const block = blockInput.getBlock();
+
+  const blobSidecars: deneb.BlobSidecars = [];
+  // response.length should always match blobMeta.length and they should be in the same order
+  for (let i = 0; i < blobMeta.length; i++) {
+    const blobAndProof = enginedResponse[i];
+
+    if (!blobAndProof) {
+      metrics?.blobs.getBlobsV1Miss.inc();
+    } else {
+      metrics?.blobs.getBlobsV1Hit.inc();
+      const {blob, proof} = blobAndProof;
+      const index = blobMeta[i].index;
+      const kzgCommitment = block.message.body.blobKzgCommitments[index];
+      const blobSidecar: deneb.BlobSidecar = {
+        index,
+        blob,
+        kzgProof: proof,
+        kzgCommitment,
+        // TODO(fulu): refactor this to only calculate the root inside these following two functions once
+        kzgCommitmentInclusionProof: computePreFuluKzgCommitmentsInclusionProof(forkName, block.message.body, index),
+        signedBlockHeader: signedBlockToSignedHeader(config, block),
+      };
+
+      blockInput.addBlob({
```

Comment:
what's happens if it already exists? based on current data by the time `getBlobsV1` resolves we will have received most/all blobs (even those we explicitly requested) through gossip

Reply:
I added a condition to check for this and metrics for both side of the conditional